python/semantic_kernel/functions/kernel_function.py

```diff
@@ -204,98 +221,139 @@ async def _non_streaming_function(*args, **kwargs):
+            )
+
+        if not prompt_template:
+            prompt_template = KernelPromptTemplate(prompt_template_config)
 
         async def _local_func(
             function: KernelFunctionMetadata,
             kernel: "Kernel",
             service: Union[TextCompletionClientBase, ChatCompletionClientBase],
             request_settings: PromptExecutionSettings,
             arguments: KernelArguments,
+            chat_history: Optional[ChatHistory] = None,
             **kwargs: Dict[str, Any],
         ) -> "FunctionResult":
             if service is None:
                 raise ValueError("AI LLM service cannot be `None`")
-            # from semantic_kernel.functions.kernel_function import KernelFunction  # noqa # pylint: disable=unused-import
 
-            # FunctionResult.model_rebuild()
+            prompt = await prompt_template.render(kernel, arguments)
+
+            if not chat_history or len(chat_history) == 0:
+                chat_history = ChatHistory(system_message=prompt)
+            else:
+                chat_history.add_user_message(prompt)
 
-            if isinstance(service, ChatCompletionClientBase):
-                try:
-                    messages = await function_config.prompt_template.render_messages(kernel, arguments)
-                    completions = await service.complete_chat(messages, request_settings)
+            try:
+                if isinstance(service, ChatCompletionClientBase):
+                    completions = await service.complete_chat(chat_history, request_settings)
                     return FunctionResult(
                         function=function,
                         value=completions,
                         metadata={
-                            "messages": messages,
+                            "messages": chat_history,
                             "arguments": arguments,
                             "metadata": [completion.metadata for completion in completions],
                         },
                     )
-                except Exception as exc:
-                    logger.error(f"Error occurred while invoking function {function.name}: {exc}")
-                    raise exc
+            except Exception as exc:
+                logger.error(f"Error occurred while invoking function {function.name}: {exc}")
+                raise exc
+
             try:
-                prompt = await function_config.prompt_template.render(kernel, arguments)
-                completions = await service.complete(prompt, request_settings)
-                return FunctionResult(
-                    function=function,
-                    value=completions,
-                    metadata={
-                        "prompt": prompt,
-                        "arguments": arguments,
-                        "metadata": [completion.metadata for completion in completions],
-                    },
-                )
+                if isinstance(service, TextCompletionClientBase):
```

Comment:
no need to check this right?

Reply:
Isn't it better to check, just in case? What if we get a random service that isn't an instance of either? Saw this in dotnet and I liked how explicit they were.

Reply:
fair enough